bpf/lib/conntrack.h

```diff
@@ -575,6 +575,7 @@ static __always_inline void ct4_cilium_dbg_tuple(struct __ctx_buff *ctx, __u8 ty
 	cilium_dbg(ctx, type, addr, rev_nat_index);
 }
 
+#ifndef __TEST__
```

Comment:
Do we have to add these wrappers around every function that we test?

It'd be nice if we could separate out the declaration vs. implementation somehow and avoid littering the code with these macros for each tested function.

Reply:
If we only mock helper functions, we can avoid that. But if developers want mock a specific customized function and not affect other functions in this file, they can wrap this function with such macros upon testing to avoid this function to be defined here (need to be defined in the mock library). We do not actually need to commit them into the tree, just a demo showing the necessary preparation before performing test.

Reply:
Could the specific test just do `#define` trickery instead in the test? Basically forward-declare their mock version of the function then `#define ct_update mock_ct_update`? Then it can be up to the individual test on how it wants to handle this, instead of requiring the developer to understand which tests mock out which functionality and conditionally enable/disable some of the functions.

Reply:
^^ This is basically what `bpf/tests/conntrack_test.h` does for the helper functions.

Reply:
I tried this approach before but failed. Since CMock generates not only a mock header but also a mock c file to be compiled together, we have no way to replace a static inline function with a mock substitute using a macro trick like conntrack_test.h does. But I am still thinking of if there is a more elegant way to mock customized functions :).

Reply:
Hmm okay, that's a bit awkward then. I don't have any other immediate suggestions, only a concern about this approach below.

> We do not actually need to commit them into the tree, just a demo showing the necessary preparation before performing test.

^ This statement seems to assume that testing is a manual process, but that should not be the case. If we require manual changes to the tree to be able to run a test, then no-one will run the test. These unit tests need to be automated such that we can run them regularly as part of CI testing pipelines. Also if a developer wants to run a test, they just run some make target locally which runs the test for them. 

Maybe this is not an immediate concern, but let's say you wanted to also unit-test `ct_lookup()`. The logic is not entirely trivial as it's not just a hash table lookup. There's IP fragment handling logic in there; there's different logic depending on the type of the packet and depending on the state in the conntrack table, etc. So I could see a desire to unit-test this logic. Separately, it looks like from this overall PR that you wanted to mock this functionality out so that you don't have to worry about bugs in ct_lookup() when testing nat.h. How can we let both tests coexist in the tree and regularly run them without manual input?

Reply:
I think these unit tests could run at different granularity/abstraction level. For example, current nat tests might focus on logic in NAT itself without worrying about bugs in `ct_lookup()` and that's why we would like to mock it out. But I can imagine we could have another testsuite that covers the logic in `ct_lookup()` itself, such a test will be compiled without using current mock header. Essentially we could have different test files with different sets of real/mock header combinations determined by what they are testing. Does this make sense?

If so then we kinda need a way to substitute these functions (e.g., `ct_lookup()`) when needed with generated mocks, not sure if there's some compiler trick we can use to achieve that purpose?

Reply:
:100: exactly my thoughts @Weil0ng . Does CMock provide some kind of functionality to help with this? It feels like this kind of "unit testing at different layers of abstraction" problem should be common across various testsuites using CMock. BPF style of defining everything in headers might make things slightly more complicated, but hopefully we can still come up with a solution.

Reply:
This problem has been solved. We can actually replace the function we want to mock like conntrack_test.h does without modifying the source file. I wrongly intrepreted an error at the first attempt. Thanks for the suggestions!